File: src/main/java/fr/umontpellier/iut/dominion/cards/component/ExtraTurnComponent.java
package fr.umontpellier.iut.dominion.cards.component;

import java.util.Optional;
import java.util.concurrent.atomic.AtomicBoolean;

public class ExtraTurnComponent implements CardComponent {
    private AtomicBoolean used;

    public ExtraTurnComponent(AtomicBoolean used) {
        this.used = used;
    }
    public Optional<ExtraTurnComponent> canUseExtraTurn() {
        return used.get()? Optional.empty(): Optional.of(this);
    }

    public void consume() {
        used.set(true);
    }

}

File: src/main/java/fr/umontpellier/iut/dominion/cards/component/ExtraTurnComponent.kt
package fr.umontpellier.iut.dominion.cards.component

import fr.umontpellier.iut.dominion.Player.Player
import java.util.concurrent.atomic.AtomicBoolean

class ExtraTurnComponent(
    private val used: AtomicBoolean = AtomicBoolean(false)
) : CardComponent {

    /**
     * Vérifie si le tour supplémentaire est disponible.
     * @return L'effet lui-même s'il est disponible, ou `null` s'il a déjà été utilisé.
     */
    fun canUseExtraTurn(): ExtraTurnComponent? {
        return if (used.get()) null else this
    }

    /**
     * Consomme le tour supplémentaire en marquant le composant comme utilisé.
     */
    fun consume(player: Player) {
        used.set(true)
    }
}
